### DISCUSSION TITLE
"append" attribute event - way to skip the append?

### DISCUSSION THREAD

**Discussion by dkbarn** [sqlalchemy/sqlalchemy#10914] [Usage Questions]
Is there any way, using either the "append" attribute event or a validator, to cause the addition of the object to the collection to be skipped? I know I can add the `retval=True` argument to the `@event.listens_for` decorator, and this will allow me to alter the value that ends up getting appended, but what if I wanted to check some criteria and decide not to add the item at all? 

I've tried returning `None` but this ends up just putting `None` in the collection and then obviously erroring.

I realize that raising an exception inside the event listener is an option but if the actual goal is to gracefully ignore the item when it fails validation, then you'd have to put a try/except around every append to that collection.

```python
@event.listens_for(SomeClass.some_attribute, 'append', retval=True)
def receive_append(target, value, initiator):
    if we_care_about(value):
        return value
    else:
        # this doesn't work
        return None
```
Is there maybe a way to "abort" the event by messing with the `initiator` or something?

**Comment by zzzeek** [comment]
not for the collections, you can only change the value that gets sent.

you'd want to build a custom collection class for this kind of thing but it seems like it would be better to not have a requirement like this for what you ultimately are building
